crates/build/src/docker.rs

```diff
@@ -0,0 +1,346 @@
+        );
+
+        let options = Some(WaitContainerOptions {
+            condition: "not-running",
+        });
+
+        let mut wait_stream = client.wait_container(&container_id, options);
+        while wait_stream.next().await.is_some() {}
+
+        client
+            .stop_container(&container_id, Some(StopContainerOptions { t: 20 }))
+            .await?;
+
+        build_steps.increment_current();
+        maybe_println!(
+            verbosity,
+            " {} {}",
+            format!("{build_steps}").bold(),
+            "Docker container has finished the build."
+                .bright_green()
+                .bold(),
+        );
+
+        let result_contents = match std::fs::read_to_string(&file_path) {
+            Ok(content) => {
+                std::fs::remove_file(&file_path)?;
+                content
+            }
+            Err(e) => {
+                std::fs::remove_file(&file_path)?;
+                anyhow::bail!(e);
+            }
+        };
+
+        let mut build_result: BuildResult = serde_json::from_str(&result_contents)
+            .map_err(|_| {
+                anyhow::anyhow!(
+                    "Error parsing output from docker build. The build probably failed!"
+                )
+            })?;
+
+        let new_path = host_folder.join(
+            build_result
+                .target_directory
+                .as_path()
+                .strip_prefix("/contract")?,
+        );
+        build_result.target_directory = new_path;
+
+        let new_path = build_result.dest_wasm.as_ref().map(|p| {
+            host_folder.join(
+                p.as_path()
+                    .strip_prefix("/contract")
+                    .expect("cannot strip prefix"),
+            )
+        });
+        build_result.dest_wasm = new_path;
+
+        build_result.metadata_result.as_mut().map(|mut m| {
+            m.dest_bundle = host_folder.join(
+                m.dest_bundle
+                    .as_path()
+                    .strip_prefix("/contract")
+                    .expect("cannot strip prefix"),
+            );
+            m.dest_metadata = host_folder.join(
+                m.dest_metadata
+                    .as_path()
+                    .strip_prefix("/contract")
+                    .expect("cannot strip prefix"),
+            );
+            m
+        });
+
+        if let Some(metadata_artifacts) = &build_result.metadata_result {
+            let mut metadata = ContractMetadata::load(&metadata_artifacts.dest_bundle)?;
+            metadata.image = Some(build_image.id.to_string());
+
+            crate::metadata::write_metadata(
+                metadata_artifacts,
+                metadata,
+                &mut build_steps,
+                &verbosity,
+            )?;
+        }
+
+        Ok(BuildResult {
+            output_type,
+            verbosity,
+            ..build_result
+        })
+    })
```

Comment:
This method is pretty large, wdyt about extracting its steps into separate functions like: `construct_cmd`, `create_container`, etc? It would make it much easier to read through and review.

Reply:
Generally speaking. I don't see where I can decompose the function since none of the blocks are resued anywhere in the codebase.

Reply:
In my mind you extract to a function not only for reuse but also to improve readability. Right now, this `docker::build` command is ~300 lines long, fully of error creation, path manipulation and mutable values - it's just difficult to follow and b/c of that review properly.

Reply:
Looks better, thanks.